### ISSUE TITLE
Using with async tape test using promises is problematic

### ISSUE THREAD

**Issue by gferreri** [sinonjs/sinon-test#75]
I have a tape test that I'm trying to wrap with sinon.test like so:
```
test('tests my new feature', sinon.test(function (t) {
   return Promise.all([promise1, promise2, promise3])
     .then(() => {
        t.end()
     })
})
```

But when I run, I get the error:

```
Error: Your test should take a callback *or* return a promise. It should not do both.
    at handleAsyncFn (/Users/gsferrer/Projects/my-api/node_modules/sinon-test/lib/test.js:53:25)
    at callSandboxedFn (/Users/gsferrer/Projects/my-api/node_modules/sinon-test/lib/test.js:90:16)
    at Test.sinonAsyncSandboxedTest (/Users/gsferrer/Projects/my-api/node_modules/sinon-test/lib/test.js:102:24)
    at Test.bound [as _cb] (/Users/gsferrer/Projects/my-api/node_modules/tape/lib/test.js:66:32)
    at Test.run (/Users/gsferrer/Projects/my-api/node_modules/tape/lib/test.js:85:10)
    at Test.bound [as run] (/Users/gsferrer/Projects/my-api/node_modules/tape/lib/test.js:66:32)
    at Immediate.next (/Users/gsferrer/Projects/my-api/node_modules/tape/lib/results.js:71:15)
    at runCallback (timers.js:649:20)
    at tryOnImmediate (timers.js:622:5)
    at processImmediate [as _immediateCallback] (timers.js:594:5)
```

Seems like my tape object is being mistaken for a callback function.

**Comment by fatso83** [comment]
Hmm ... good catch. The code for this module is _very_ small. Care to take a stab at patching it? I am sure it is not much more than a `typeof arg === function` check that is needed somewhere.

**Comment by zixia** [comment]
I ran into this issue today.

The following code will use `handleAsyncFn` if we pass `function (t) {}` to it, because `t` is a parameter makes `callback.length` to be true.

https://github.com/sinonjs/sinon-test/blob/931ab831fb08c11a098b1f7f1be0e05b840c121b/lib/test.js#L102-L104

However, `t` is not a callback but a parameter from `tape`(or `tap`, which I'm using).

And if we return a promise, we got the problem at here:

https://github.com/sinonjs/sinon-test/blob/931ab831fb08c11a098b1f7f1be0e05b840c121b/lib/test.js#L51-L57

My code to reproduce this issue is the following code:

```ts
t.test('test', sinonTest(async function (t) {
  t.pass('ok')
})
```
